x/mongo/driver/auth/creds/awscreds.go

```diff
@@ -0,0 +1,187 @@
+	AccessKeyID     string `json:"AccessKeyId"`
+	SecretAccessKey string `json:"SecretAccessKey"`
+	Token           string `json:"Token"`
+}
+
+const (
+	awsRelativeURI     = "http://169.254.170.2/"
+	awsEC2URI          = "http://169.254.169.254/"
+	awsEC2RolePath     = "latest/meta-data/iam/security-credentials/"
+	awsEC2TokenPath    = "latest/api/token"
+	defaultHTTPTimeout = 10 * time.Second
+)
+
+// AwsCredentials contains AWS credential fields.
+type AwsCredentials struct {
+	Username string
+	Password string
+	Token    string
+}
+
+// ValidateAndMakeCredentials validates credential fields and packs them into awsv4.StaticProvider.
+func (ac AwsCredentials) ValidateAndMakeCredentials() (*awsv4.StaticProvider, error) {
+	if ac.Username != "" && ac.Password == "" {
+		return nil, errors.New("ACCESS_KEY_ID is set, but SECRET_ACCESS_KEY is missing")
+	}
+	if ac.Username == "" && ac.Password != "" {
+		return nil, errors.New("SECRET_ACCESS_KEY is set, but ACCESS_KEY_ID is missing")
+	}
+	if ac.Username == "" && ac.Password == "" && ac.Token != "" {
+		return nil, errors.New("AWS_SESSION_TOKEN is set, but ACCESS_KEY_ID and SECRET_ACCESS_KEY are missing")
+	}
+	if ac.Username != "" || ac.Password != "" || ac.Token != "" {
+		return &awsv4.StaticProvider{Value: awsv4.Value{
+			AccessKeyID:     ac.Username,
+			SecretAccessKey: ac.Password,
+			SessionToken:    ac.Token,
+		}}, nil
+	}
+	return nil, nil
+}
+
+// AwsCredentialProvider provides AWS credentials.
+type AwsCredentialProvider struct {
+	HTTPClient *http.Client
+}
+
+func executeAWSHTTPRequest(ctx context.Context, httpClient *http.Client, req *http.Request) ([]byte, error) {
+	ctx, cancel := context.WithTimeout(ctx, defaultHTTPTimeout)
+	defer cancel()
+	resp, err := httpClient.Do(req.WithContext(ctx))
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	return ioutil.ReadAll(resp.Body)
+}
+
+// getEC2Credentials generates EC2 credentials.
+func (p *AwsCredentialProvider) getEC2Credentials(ctx context.Context) (*awsv4.StaticProvider, error) {
+	// get token
+	req, err := http.NewRequest("PUT", awsEC2URI+awsEC2TokenPath, nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("X-aws-ec2-metadata-token-ttl-seconds", "30")
+
+	token, err := executeAWSHTTPRequest(ctx, p.HTTPClient, req)
+	if err != nil {
+		return nil, err
+	}
+	if len(token) == 0 {
+		return nil, errors.New("unable to retrieve token from EC2 metadata")
+	}
+	tokenStr := string(token)
+
+	// get role name
+	req, err = http.NewRequest("GET", awsEC2URI+awsEC2RolePath, nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("X-aws-ec2-metadata-token", tokenStr)
+
+	role, err := executeAWSHTTPRequest(ctx, p.HTTPClient, req)
+	if err != nil {
+		return nil, err
+	}
+	if len(role) == 0 {
+		return nil, errors.New("unable to retrieve role_name from EC2 metadata")
+	}
+
+	// get credentials
+	pathWithRole := awsEC2URI + awsEC2RolePath + string(role)
+	req, err = http.NewRequest("GET", pathWithRole, nil)
```

Comment:
```suggestion
	req, err = http.NewRequest(http.MethodGet, pathWithRole, nil)
```

Suggest using the [HTTP constants](https://pkg.go.dev/net/http#MethodPut).